### PR TITLE
Large ti db fixes

### PR DESCRIPTION
**Pull request by blauwers**
Required items, please complete
   
   Change(s):
  - Updated .script/tests/KqlvalidationsTests/KqlValidationTests.cs
  - Updated .script/tests/KqlvalidationsTests/SkipValidationsTemplates.json
  - Added .script/tests/KqlvalidationsTests/CustomTables/UrlClickEvents.json
  - Updated Solutions/Threat Intelligence/Analytic Rules/DomainEntity_CommonSecurityLog.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/DomainEntity_DnsEvents.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/DomainEntity_PaloAlto.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/DomainEntity_SecurityAlert.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/DomainEntity_Syslog.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/DomainEntity_imWebSession.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/EmailEntity_AzureActivity.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/EmailEntity_OfficeActivity.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/EmailEntity_PaloAlto.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/EmailEntity_SecurityAlert.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/EmailEntity_SecurityEvent.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/EmailEntity_SigninLogs.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/FileHashEntity_CommonSecurityLog.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/FileHashEntity_SecurityEvent.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/IPEntity_AWSCloudTrail.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/IPEntity_AppServiceHTTPLogs.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/IPEntity_AzureActivity.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/IPEntity_AzureFirewall.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/IPEntity_AzureKeyVault.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/IPEntity_AzureNetworkAnalytics.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/IPEntity_AzureSQL.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/IPEntity_CustomSecurityLog.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/IPEntity_DnsEvents.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/IPEntity_DuoSecurity.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/IPEntity_OfficeActivity.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/IPEntity_SigninLogs.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/IPEntity_VMConnection.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/IPEntity_W3CIISLog.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/IPEntity_imNetworkSession.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/IPEntity_imWebSession.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/Threat Intel Matches to GitHub Audit Logs.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/URLEntity_AuditLogs.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/URLEntity_OfficeActivity.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/URLEntity_PaloAlto.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/URLEntity_SecurityAlerts.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/URLEntity_Syslog.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/imDns_DomainEntity_DnsEvents.yaml
  - Updated Solutions/Threat Intelligence/Analytic Rules/imDns_IPEntity_DnsEvents.yaml
  - Added Solutions/Threat Intelligence/Analytic Rules/DomainEntity_DeviceNetworkEvents.yaml
  - Added Solutions/Threat Intelligence/Analytic Rules/DomainEntity_EmailEvents.yaml
  - Added Solutions/Threat Intelligence/Analytic Rules/DomainEntity_EmailUrlInfo.yaml
  - Added Solutions/Threat Intelligence/Analytic Rules/FileHashEntity_DeviceFileEvents.yaml
  - Added Solutions/Threat Intelligence/Analytic Rules/EmailEntity_EmailEvents.yaml
  - Added Solutions/Threat Intelligence/Analytic Rules/IPEntity_DeviceNetworkEvents.yaml
  - Added Solutions/Threat Intelligence/Analytic Rules/UrlEntityEntity_DeviceNetworkEvents.yaml
  - Added Solutions/Threat Intelligence/Analytic Rules/URLEntity_EmailUrlInfo.yaml
  - Added Solutions/Threat Intelligence/Analytic Rules/URLEntity_UrlClickEvents.yaml
  - Updated Solutions/Threat Intelligence/Data/Solution_ThreatIntelligenceTemplateSpec.json
  - Updated Solutions/Threat Intelligence/Package/createUiDefinition.json
  - Updated Solutions/Threat Intelligence/Package/mainTemplate.json
  - Updated Solutions/Microsoft Defender XDR/Data Connectors/MicrosoftThreatProtection.JSON
  - Updated Solutions/Microsoft Defender XDR/Data/Solution_Microsoft Defender XDR.json

   Reason for Change(s):
   - Implement more robust prefiltering of data to optimize query and allow large TI database (>20 million active indicators) without timeout
   - Fix TI query validation test to allow filtering per KQL best practices
   - New analytic queries to enhance phishing detection
   - Declare UrlClickEvents in MicrosoftThreatProtection
   - Declare UrlClickEvents table for automated tests
   - Merge changes from PR#9181 by @kfriede
 
   Version Updated:
   - Yes

   Testing Completed:
   - Yes

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes

### CODE REVIEW

**Comment by blauwers** [comment]
@microsoft-github-policy-service agree

**Comment by blauwers** [comment]
Sorry for all the commits - it seems all the validation tests are now working.

**Comment by v-rbajaj** [comment]
Hi @blauwers, please update `version` of `Solutions/Threat Intelligence/Analytic Rules/IPEntity_AzureKeyVault.yaml`?

**Comment by v-rbajaj** [comment]
Hi @blauwers, there is another PR #https://github.com/Azure/Azure-Sentinel/pull/9181/files opened for Threat intelligence and there are few contents which are similar. 

We would need to repackage this solution after the other PR is merged so that it doesn't raise the conflict.

**Comment by blauwers** [comment]
@v-rbajaj I would be happy to port over relevant changes to this PR if that is helpful.

**Comment by v-rbajaj** [comment]
Hi @blauwers, there are couple of challenges, we can't simply close that PR unless the other author agrees to it. 

And if you package this solution with 3.0.2 version than the other PR will have merge conflicts. 

Can we wait for the other PR to close and then work on this ?

**Comment by blauwers** [comment]
Hello @v-rbajaj,

I am used to seeing code that passes the QA test going before code that does not. That said, the code in the other PR puts some things out of order in such a way it will break the accuracy of the TI queries. Specifically, the statement `| where Active == true and ExpirationDateTime > now()` needs to come after `| summarize [.]`

This PR also fixes the quality test that previously made analytic TI queries fail to validate - ergo why the tests are now passing, which is not addressed in the other PR.

It would be faster to merge or port the changes from the other PR to this PR and fix the query issues simultaneously. Can we agree that this is an acceptable strategy? If so, I will start working on merging the PRs and making the necessary corrections.

**Comment by blauwers** [comment]
To illustrate the point:

```
let table = datatable (TimeGenerated: datetime , Active: bool, ExpirationDateTime: datetime, IndicatorId: string, Description: string) [
 datetime(2023-11-17 7:00 AM), true, datetime(2023-11-20 9:00 AM), "Indicator1", "Active indicator 1 from Source 1",
 datetime(2023-11-18 7:00 AM), false, datetime(2023-11-20 9:00 AM), "Indicator1", "Inctive indicator 1 from Source 1",
 datetime(2023-11-19 7:00 AM), true, datetime(2023-11-19 9:00 AM), "Indicator2", "Active indicator 1 from Source 2",
];
table
| where TimeGenerated >= datetime(2023-11-10 9:00 AM)
| summarize LatestIndicatorTime = arg_max(TimeGenerated, *) by IndicatorId
| where Active == true and ExpirationDateTime > datetime(2023-11-19 7:00 AM)
```

outputs something else than the following

```
let table = datatable (TimeGenerated: datetime , Active: bool, ExpirationDateTime: datetime, IndicatorId: string, Description: string) [
 datetime(2023-11-17 7:00 AM), true, datetime(2023-11-20 9:00 AM), "Indicator1", "Active indicator 1 from Source 1",
 datetime(2023-11-18 7:00 AM), false, datetime(2023-11-20 9:00 AM), "Indicator1", "Inctive indicator 1 from Source 1",
 datetime(2023-11-19 7:00 AM), true, datetime(2023-11-19 9:00 AM), "Indicator2", "Active indicator 1 from Source 2",
];
table
| where TimeGenerated >= datetime(2023-11-10 9:00 AM)
| where Active == true and ExpirationDateTime > datetime(2023-11-19 7:00 AM)
| summarize LatestIndicatorTime = arg_max(TimeGenerated, *) by IndicatorId
```

As you can see the latter causes false positives to show up.

**Comment by kfriede** [comment]
@blauwers if you want to pull the latest changes in from #9181 again, I'd be okay with @v-rbajaj approving this in place and I can close #9181

**Comment by blauwers** [comment]
Well that was entirely messed up and I will start over I guess. @kfriede after I pulled your latest PR I suddenly had 400+ files that had changed. :(

**Comment by v-rbajaj** [comment]
Hi @blauwers, thanks for making these changes, will check and get back to you by 27 Nov 2023.

File: Solutions/Threat Intelligence/Package/createUiDefinition.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
     "config": {
       "isWizard": false,
       "basics": {
-        "description": "<img src=\"https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/Logos/Azure_Sentinel.svg\"width=\"75px\" height=\"75px\">\n\n**Note:** Please refer to the following before installing the solution: \r \n • Review the solution [Release Notes](https://github.com/Azure/Azure-Sentinel/tree/master/Solutions/Threat Intelligence/ReleaseNotes.md)\r \n • There may be [known issues](https://aka.ms/sentinelsolutionsknownissues) pertaining to this Solution, please refer to them before installing.\n\nThe Threat Intelligence solution contains data connectors for import of threat indicators into Microsoft Sentinel, analytic rules for matching TI data with event data, workbook, and hunting queries. Threat indicators can be malicious IP's, URL's, filehashes, domains, email addresses etc.\n\n**Data Connectors:** 4, **Workbooks:** 1, **Analytic Rules:** 38, **Hunting Queries:** 5\n\n[Learn more about Microsoft Sentinel](https://aka.ms/azuresentinel) | [Learn more about Solutions](https://aka.ms/azuresentinelsolutionsdoc)",
```
Comment (v-rbajaj) [inline Solutions/Threat Intelligence/Package/createUiDefinition.json:9]: Hi @blauwers, can you please revert this change and update the create UI in the 3.0.2 zip as well?

Comment (blauwers) [reply Solutions/Threat Intelligence/Package/createUiDefinition.json:9]: Hello @v-rbajaj This change comes from the V3 packaging tool. I reverted that change and ended up having to create the 3.0.2.zip by hand.

**Comment by v-rbajaj** [comment]
Hi @blauwers, thanks for the changes. I was going through the PR and I noticed that you have made changes in Microsoft Defender XDR's Data Connector, so for that we would need you to repackage Microsoft Defender XDR using v3 tool. 

Apart from that I am reviewing the changes in Threat Intelligence.

**Comment by blauwers** [comment]
Hello @v-rbajaj, I updated the package as requested.

**Comment by v-rbajaj** [comment]
Hi @blauwers, please resolve merge conflicts from the PR.

**Comment by blauwers** [comment]
Hell @v-rbajaj as requested, I have resolved the merge conflicts by merging my branch with an up to date master.

**Comment by blauwers** [comment]
@v-rbajaj I made the requested changes.

While doing so, I noticed I must not have re-merged the version bump for Microsoft Defender XDR, so I updated that to 3.0.2 again. Thank you for pointing out the issue with the empty arrays causing the ARM TTK to fail. How many issues have been in files generated by a packaging tool is odd.

File: Solutions/Threat Intelligence/Package/createUiDefinition.json [hunk 2]
```diff
@@ -6,7 +6,7 @@
...
+        "description": "<img src=\"https://raw.githubusercontent.com/Azure/Azure-Sentinel/master/Logos/Azure_Sentinel.svg\"width=\"75px\" height=\"75px\">\n\n**Note:** Please refer to the following before installing the solution: \r \n • Review the solution [Release Notes](https://github.com/Azure/Azure-Sentinel/tree/master/Solutions/Threat Intelligence/ReleaseNotes.md)\r \n • There may be [known issues](https://aka.ms/sentinelsolutionsknownissues) pertaining to this Solution, please refer to them before installing.\n\nThe Threat Intelligence solution contains data connectors for import of threat indicators into Microsoft Sentinel, analytic rules for matching TI data with event data, workbook, and hunting queries. Threat indicators can be malicious IP's, URL's, filehashes, domains, email addresses etc.\n\n**Data Connectors:** 4, **Workbooks:** 1, **Analytic Rules:** 44, **Hunting Queries:** 5\n\n[Learn more about Microsoft Sentinel](https://aka.ms/azuresentinel) | [Learn more about Solutions](https://aka.ms/azuresentinelsolutionsdoc)",
```
Comment (v-rbajaj) [inline Solutions/Threat Intelligence/Package/createUiDefinition.json:9]: Hi @blauwers, after adding new analytical rules the count of analytical rules should be 47 instead of 44. 
Please update the count here  `**Analytic Rules:** 44` to 47

After doing so, you need to update the 3.0.2 zip create ui.

File: Solutions/Threat Intelligence/Package/createUiDefinition.json [hunk 3]
```diff
@@ -216,529 +167,655 @@
...
             "type": "Microsoft.Common.Section",
-            "label": "TI map Domain entity to Syslog",
+            "label": "TI map Domain entity to Web Session Events (ASIM Web Session schema)",
             "elements": [
               {
                 "name": "analytic6-text",
                 "type": "Microsoft.Common.TextBlock",
                 "options": {
-                  "text": "Identifies a match in Syslog table from any Domain IOC from TI"
+                  "text": "This rule identifies Web Sessions for which the target URL hostname is a known IoC. <br><br>This rule uses the [Advanced Security Information Model (ASIM)](https:/aka.ms/AboutASIM) and supports any web session source that complies with ASIM."
```
Comment (v-rbajaj) [inline Solutions/Threat Intelligence/Package/createUiDefinition.json:246]: Please remove these HTML tags <br><br>

File: Solutions/Threat Intelligence/Package/createUiDefinition.json (hunk 2 above)
Comment (blauwers) [reply Solutions/Threat Intelligence/Package/createUiDefinition.json]: Got it. Since that comes from the packaging tool, I forgot to update it after putting the preferred description back. Apologies.

File: Solutions/Threat Intelligence/Package/createUiDefinition.json (hunk 3 above)
Comment (blauwers) [reply Solutions/Threat Intelligence/Package/createUiDefinition.json]: I removed the HTML tags from all their origins in the solution.

File: Solutions/Microsoft Defender XDR/Package/createUiDefinition.json [hunk 4]
```diff
@@ -6,7 +6,7 @@
...
...n for Microsoft Sentinel enables you to ingest Security Alerts/Incidents and raw logs from the products within Microsoft Defender XDR suite into Microsoft Sentinel.\n\nAdditional Hunting Queries to support proactive and reactive hunting for the Microsoft Defender XDR solution can be found on [GitHub](https://github.com/Azure/Azure-Sentinel/tree/master/Hunting%20Queries/Microsoft%20365%20Defender). This repository has a collection of queries developed by Microsoft Security Research and Microsoft Sentinel community contributions.\n\n**Underlying Microsoft Technologies used:**\n\nThis solution takes a dependency on the following technologies, and some of these dependencies either may be in [Preview](https://azure.microsoft.com/support/legal/preview-supplemental-terms/) state or might result in additional ingestion or operational costs:\n\na. [Azure Monitor HTTP Data Collector API](https://docs.microsoft.com/azure/azure-monitor/logs/data-collector-api)\n\n**Data Connectors:** 1, **Workbooks:** 3, **Analytic Rules:** 10, **Hunting Queries:** 4\n\n[Learn more about Microsoft Sentinel](https://aka.ms/azuresentinel) | [Learn more about Solutions](https://aka.ms/azuresentinelsolutionsdoc)",
```
Comment (v-rbajaj) [inline Solutions/Microsoft Defender XDR/Package/createUiDefinition.json:9]: Hi @blauwers, everything looks fine in this PR, only thing is you need to revert this create ui change and update it 3.0.2 zip. 

Once you are done with this, we are good to merge this PR.

**Review by v-rbajaj — APPROVED** [review]

**Review by v-dvedak — APPROVED** [review]
